src/mapgen_functions.cpp

```diff
@@ -977,382 +980,319 @@ void mapgen_fungal_flowers(map *m, oter_id, mapgendata dat, int, float)
 }
 
-void mapgen_road_straight(map *m, oter_id terrain_type, mapgendata dat, int turn, float density)
-{
-    bool sidewalks = false;
-    for (int i = 0; i < 8; i++) {
-        if (otermap[dat.t_nesw[i]].has_flag(has_sidewalk)) {
-            sidewalks = true;
+int terrain_type_with_suffix_to_nesw_array(oter_id terrain_type, bool array[4]) {
+    // extract the suffix from the terrain type name
+    std::string suffix = std::string(terrain_type).substr( std::string(terrain_type).find_last_of("_") + 1 );
+    // non-"_end" end tiles have _north _east _south _west, all of which contain "t"
+    if (suffix.find("t") != std::string::npos) {
+        suffix = suffix.substr(0,1);
+        if      (suffix=="n") suffix = "s"; // and they are all backwards :(
+        else if (suffix=="e") suffix = "w";
+        else if (suffix=="s") suffix = "n";
+        else if (suffix=="w") suffix = "e";
+    }
+    // manhole exception
+    if (suffix=="manhole") suffix="nesw";
+    // count and mark which directions the road goes
+    int num_dirs = 0;
+    if (suffix.find("n") != std::string::npos) { array[0] = true; num_dirs++; } else { array[0] = false; }
+    if (suffix.find("e") != std::string::npos) { array[1] = true; num_dirs++; } else { array[1] = false; }
+    if (suffix.find("s") != std::string::npos) { array[2] = true; num_dirs++; } else { array[2] = false; }
+    if (suffix.find("w") != std::string::npos) { array[3] = true; num_dirs++; } else { array[3] = false; }
+    return num_dirs;
+}
+
+// perform dist counterclockwise rotations on a nesw or neswx array
+template<typename T> void nesw_array_rotate(T *array, size_t len, size_t dist) {
+    if( len == 4 ) {
+        for (;dist--;) {
+            T temp = array[0];
+            array[0] = array[1];
+            array[1] = array[2];
+            array[2] = array[3];
+            array[3] = temp;
         }
-    }
-
-    VehicleSpawn::apply((sidewalks) ? vspawn_id("default_city") : vspawn_id("default_country"), *m, "road_straight");
-
-    for (int i = 0; i < SEEX * 2; i++) {
-        for (int j = 0; j < SEEY * 2; j++) {
-            if (i < 4 || i >= SEEX * 2 - 4) {
-                if (sidewalks) {
-                    m->ter_set(i, j, t_sidewalk);
-                } else {
-                    m->ter_set(i, j, dat.groundcover());
-                }
-            } else {
-                if ((i == SEEX - 1 || i == SEEX) && j % 4 != 0) {
-                    m->ter_set(i, j, t_pavement_y);
-                } else {
-                    m->ter_set(i, j, t_pavement);
-                }
-            }
-        }
-    }
-    if (terrain_type == "road_ew") {
-        m->rotate(1);
-    }
-    if(sidewalks) {
-        m->place_spawns( mongroup_id( "GROUP_ZOMBIE" ), 2, 0, 0, SEEX * 2 - 1, SEEX * 2 - 1, density);
-        // 1 per 10 overmaps
-        if (one_in(10000)) {
-            m->add_spawn(mon_zombie_jackson, 1, SEEX, SEEY);
+    } else {
+        for (;dist--;) {
```

Comment:
`for(;n--;)` is clean and obvious to me... I guess that's a style decision. If you really want `for(int i=0;i<n;i++)` I could do that.

In retrospect, this was only a for() at all because it came from another loop that had more expressions in there. It should be `while(dist--)`

I'll fix all the if(int) instances, although some will be `!=0` since they can be negative.